diff -up gnutls-2.0.4/doc/examples/ex-serv-srp.c.nosrp gnutls-2.0.4/doc/examples/ex-serv-srp.c
--- gnutls-2.0.4/doc/examples/ex-serv-srp.c.nosrp	2007-11-15 16:24:56.000000000 +0100
+++ gnutls-2.0.4/doc/examples/ex-serv-srp.c	2007-11-29 14:50:37.000000000 +0100
@@ -29,6 +29,7 @@
 #define MAX_BUF 1024
 #define PORT 5556		/* listen to 5556 port */
 
+#if 0
 /* These are global */
 gnutls_srp_server_credentials_t srp_cred;
 gnutls_certificate_credentials_t cert_cred;
@@ -57,10 +58,12 @@ initialize_tls_session (void)
 
   return session;
 }
+#endif
 
 int
 main (void)
 {
+#if 0
   int err, listen_sd, i;
   int sd, ret;
   struct sockaddr_in sa_serv;
@@ -173,7 +176,7 @@ main (void)
   gnutls_certificate_free_credentials (cert_cred);
 
   gnutls_global_deinit ();
-
+#endif
   return 0;
 
 }
diff -up gnutls-2.0.4/doc/examples/ex-session-info.c.nosrp gnutls-2.0.4/doc/examples/ex-session-info.c
--- gnutls-2.0.4/doc/examples/ex-session-info.c.nosrp	2007-11-15 16:24:56.000000000 +0100
+++ gnutls-2.0.4/doc/examples/ex-session-info.c	2007-11-29 15:54:12.000000000 +0100
@@ -35,12 +35,12 @@ print_info (gnutls_session_t session)
       printf ("- TLS/IA session\n");
       break;
 
-
+#if 0
     case GNUTLS_CRD_SRP:
       printf ("- SRP session with username %s\n",
 	      gnutls_srp_server_get_username (session));
       break;
-
+#endif
     case GNUTLS_CRD_PSK:
       if (gnutls_psk_server_get_username (session) != NULL)
 	printf ("- PSK authentication. Connected as '%s'\n",
diff -up gnutls-2.0.4/doc/examples/ex-client-srp.c.nosrp gnutls-2.0.4/doc/examples/ex-client-srp.c
--- gnutls-2.0.4/doc/examples/ex-client-srp.c.nosrp	2007-11-15 16:24:56.000000000 +0100
+++ gnutls-2.0.4/doc/examples/ex-client-srp.c	2007-11-29 14:50:37.000000000 +0100
@@ -24,6 +24,7 @@ extern void tcp_close (int sd);
 int
 main (void)
 {
+#if 0
   int ret;
   int sd, ii;
   gnutls_session_t session;
@@ -121,6 +122,6 @@ end:
   gnutls_certificate_free_credentials (cert_cred);
 
   gnutls_global_deinit ();
-
+#endif
   return 0;
 }
