--- gnutls-1.2.9/doc/examples/ex-client-srp.c.nosrp	2005-08-10 11:13:30.000000000 +0200
+++ gnutls-1.2.9/doc/examples/ex-client-srp.c	2005-11-23 11:58:46.000000000 +0100
@@ -17,13 +17,16 @@
 #define SA struct sockaddr
 #define MSG "GET / HTTP/1.0\r\n\r\n"
 
+#if 0
 const int kx_priority[] = { GNUTLS_KX_SRP, GNUTLS_KX_SRP_DSS,
   GNUTLS_KX_SRP_RSA, 0
 };
+#endif
 
 int
 main (void)
 {
+#if 0
   int ret;
   int sd, ii;
   gnutls_session_t session;
@@ -122,6 +125,6 @@
   gnutls_certificate_free_credentials (cert_cred);
 
   gnutls_global_deinit ();
-
+#endif
   return 0;
 }
--- gnutls-1.2.9/doc/examples/ex-serv-srp.c.nosrp	2005-09-23 00:40:19.000000000 +0200
+++ gnutls-1.2.9/doc/examples/ex-serv-srp.c	2005-11-23 12:02:44.000000000 +0100
@@ -26,6 +26,7 @@
 #define MAX_BUF 1024
 #define PORT 5556		/* listen to 5556 port */
 
+#if 0
 /* These are global */
 gnutls_srp_server_credentials_t srp_cred;
 gnutls_certificate_credentials_t cert_cred;
@@ -54,10 +55,12 @@
 
   return session;
 }
+#endif
 
 int
 main (void)
 {
+#if 0
   int err, listen_sd, i;
   int sd, ret;
   struct sockaddr_in sa_serv;
@@ -170,7 +173,7 @@
   gnutls_certificate_free_credentials (cert_cred);
 
   gnutls_global_deinit ();
-
+#endif
   return 0;
 
 }
