--- a/lib/fips.c	2020-01-01 21:10:19.000000000 +0100
+++ b/lib/fips.c	2020-05-13 17:29:43.098868100 +0200
@@ -136,8 +136,8 @@
 }
 
 #define GNUTLS_LIBRARY_NAME "libgnutls.so.30"
-#define NETTLE_LIBRARY_NAME "libnettle.so.6"
-#define HOGWEED_LIBRARY_NAME "libhogweed.so.4"
+#define NETTLE_LIBRARY_NAME "libnettle.so.8"
+#define HOGWEED_LIBRARY_NAME "libhogweed.so.6"
 #define GMP_LIBRARY_NAME "libgmp.so.10"
 
 #define HMAC_SUFFIX ".hmac"
