diff -up gnutls-2.8.5/doc/examples/Makefile.am.ImplicitDSOLinking_libgcrypt gnutls-2.8.5/doc/examples/Makefile.am
--- gnutls-2.8.5/doc/examples/Makefile.am.ImplicitDSOLinking_libgcrypt	2009-06-02 13:59:32.000000000 -0500
+++ gnutls-2.8.5/doc/examples/Makefile.am	2010-02-15 19:18:55.722180984 -0600
@@ -30,7 +30,7 @@ LDADD = libexamples.la				\
 	../../lib/libgnutls.la			\
 	../../libextra/libgnutls-extra.la	\
 	../../gl/libgnu.la			\
-	$(LIBSOCKET)
+	$(LTLIBGCRYPT) $(LIBSOCKET)
 
 CXX_LDADD = $(LDADD) \
 	../../lib/libgnutlsxx.la
diff -up gnutls-2.8.5/doc/examples/Makefile.in.ImplicitDSOLinking_libgcrypt gnutls-2.8.5/doc/examples/Makefile.in
--- gnutls-2.8.5/doc/examples/Makefile.in.ImplicitDSOLinking_libgcrypt	2009-11-02 04:36:00.000000000 -0600
+++ gnutls-2.8.5/doc/examples/Makefile.in	2010-02-15 19:19:46.900440435 -0600
@@ -810,7 +810,7 @@ LDADD = libexamples.la				\
 	../../lib/libgnutls.la			\
 	../../libextra/libgnutls-extra.la	\
 	../../gl/libgnu.la			\
-	$(LIBSOCKET)
+	$(LTLIBGCRYPT) $(LIBSOCKET)
 
 CXX_LDADD = $(LDADD) \
 	../../lib/libgnutlsxx.la
diff -up gnutls-2.8.5/src/Makefile.am.ImplicitDSOLinking_libgcrypt gnutls-2.8.5/src/Makefile.am
--- gnutls-2.8.5/src/Makefile.am.ImplicitDSOLinking_libgcrypt	2009-06-02 13:59:32.000000000 -0500
+++ gnutls-2.8.5/src/Makefile.am	2010-02-15 19:22:31.674183953 -0600
@@ -40,7 +40,7 @@ noinst_LTLIBRARIES =
 
 gnutls_serv_SOURCES = list.h serv.c common.h common.c certtool-common.h
 gnutls_serv_LDADD = ../lib/libgnutls.la ../libextra/libgnutls-extra.la
-gnutls_serv_LDADD += libcmd-serv.la ../gl/libgnu.la $(LIBSOCKET) $(GETADDRINFO_LIB)
+gnutls_serv_LDADD += libcmd-serv.la ../gl/libgnu.la $(LTLIBGCRYPT) $(LIBSOCKET) $(GETADDRINFO_LIB)
 noinst_LTLIBRARIES += libcmd-serv.la
 libcmd_serv_la_CFLAGS =
 libcmd_serv_la_SOURCES = serv.gaa serv-gaa.h serv-gaa.c
@@ -61,14 +61,14 @@ libcmd_psk_la_SOURCES = psk.gaa psk-gaa.
 
 gnutls_cli_SOURCES = cli.c common.h common.c
 gnutls_cli_LDADD = ../lib/libgnutls.la ../libextra/libgnutls-extra.la
-gnutls_cli_LDADD += libcmd-cli.la ../gl/libgnu.la $(LIBSOCKET) $(GETADDRINFO_LIB)
+gnutls_cli_LDADD += libcmd-cli.la ../gl/libgnu.la $(LTLIBGCRYPT) $(LIBSOCKET) $(GETADDRINFO_LIB)
 noinst_LTLIBRARIES += libcmd-cli.la
 libcmd_cli_la_CFLAGS =
 libcmd_cli_la_SOURCES = cli.gaa cli-gaa.h cli-gaa.c
 
 gnutls_cli_debug_SOURCES = tls_test.c tests.h tests.c common.h common.c
 gnutls_cli_debug_LDADD = ../lib/libgnutls.la libcmd-cli-debug.la
-gnutls_cli_debug_LDADD += ../gl/libgnu.la $(LIBSOCKET) $(GETADDRINFO_LIB)
+gnutls_cli_debug_LDADD += ../gl/libgnu.la $(LTLIBGCRYPT) $(LIBSOCKET) $(GETADDRINFO_LIB)
 noinst_LTLIBRARIES += libcmd-cli-debug.la
 libcmd_cli_debug_la_CFLAGS =
 libcmd_cli_debug_la_SOURCES = tls_test.gaa tls_test-gaa.h tls_test-gaa.c
diff -up gnutls-2.8.5/src/Makefile.in.ImplicitDSOLinking_libgcrypt gnutls-2.8.5/src/Makefile.in
--- gnutls-2.8.5/src/Makefile.in.ImplicitDSOLinking_libgcrypt	2010-02-15 19:19:46.894442396 -0600
+++ gnutls-2.8.5/src/Makefile.in	2010-02-15 19:22:34.146181051 -0600
@@ -827,7 +827,7 @@ noinst_LTLIBRARIES = libcmd-serv.la $(am
 	libcmd-certtool.la
 gnutls_serv_SOURCES = list.h serv.c common.h common.c certtool-common.h
 gnutls_serv_LDADD = ../lib/libgnutls.la ../libextra/libgnutls-extra.la \
-	libcmd-serv.la ../gl/libgnu.la $(LIBSOCKET) $(GETADDRINFO_LIB)
+	libcmd-serv.la ../gl/libgnu.la $(LTLIBGCRYPT) $(LIBSOCKET) $(GETADDRINFO_LIB)
 libcmd_serv_la_CFLAGS = 
 libcmd_serv_la_SOURCES = serv.gaa serv-gaa.h serv-gaa.c
 @ENABLE_SRP_TRUE@srptool_SOURCES = crypt.c
@@ -840,12 +840,12 @@ libcmd_psk_la_CFLAGS = 
 libcmd_psk_la_SOURCES = psk.gaa psk-gaa.h psk-gaa.c
 gnutls_cli_SOURCES = cli.c common.h common.c
 gnutls_cli_LDADD = ../lib/libgnutls.la ../libextra/libgnutls-extra.la \
-	libcmd-cli.la ../gl/libgnu.la $(LIBSOCKET) $(GETADDRINFO_LIB)
+	libcmd-cli.la ../gl/libgnu.la $(LTLIBGCRYPT) $(LIBSOCKET) $(GETADDRINFO_LIB)
 libcmd_cli_la_CFLAGS = 
 libcmd_cli_la_SOURCES = cli.gaa cli-gaa.h cli-gaa.c
 gnutls_cli_debug_SOURCES = tls_test.c tests.h tests.c common.h common.c
 gnutls_cli_debug_LDADD = ../lib/libgnutls.la libcmd-cli-debug.la \
-	../gl/libgnu.la $(LIBSOCKET) $(GETADDRINFO_LIB)
+	../gl/libgnu.la $(LTLIBGCRYPT) $(LIBSOCKET) $(GETADDRINFO_LIB)
 libcmd_cli_debug_la_CFLAGS = 
 libcmd_cli_debug_la_SOURCES = tls_test.gaa tls_test-gaa.h tls_test-gaa.c
 certtool_SOURCES = certtool.c prime.c
