diff -up ./tests/gnutls-cli-debug.sh.gost ./tests/gnutls-cli-debug.sh
--- ./tests/gnutls-cli-debug.sh.gost	2021-02-09 13:28:46.528821113 +0100
+++ ./tests/gnutls-cli-debug.sh	2021-02-09 13:29:18.851646678 +0100
@@ -217,6 +217,8 @@ if test "${ENABLE_GOST}" = "1" && test "
 	kill ${PID}
 	wait
 
+	cat $OUTFILE
+
 	check_text "for VKO GOST-2012 (draft-smyshlyaev-tls12-gost-suites) support... yes"
 	check_text "for GOST28147-CNT cipher (draft-smyshlyaev-tls12-gost-suites) support... yes"
 	check_text "for GOST28147-IMIT MAC (draft-smyshlyaev-tls12-gost-suites) support... yes"
