diff -up gnutls-3.1.10/tests/dsa/testdsa.rndport gnutls-3.1.10/tests/dsa/testdsa
--- gnutls-3.1.10/tests/dsa/testdsa.rndport	2013-03-21 21:42:28.000000000 +0100
+++ gnutls-3.1.10/tests/dsa/testdsa	2013-03-26 15:50:38.973561603 +0100
@@ -23,7 +23,7 @@
 srcdir="${srcdir:-.}"
 SERV="${SERV:-../../src/gnutls-serv} -q"
 CLI="${CLI:-../../src/gnutls-cli}"
-PORT="${PORT:-5559}"
+PORT="${PORT:-$((RANDOM+5000))}"
 DEBUG=""
 unset RETCODE
 
diff -up gnutls-3.1.10/tests/openpgp-certs/testcerts.rndport gnutls-3.1.10/tests/openpgp-certs/testcerts
--- gnutls-3.1.10/tests/openpgp-certs/testcerts.rndport	2013-03-21 21:42:28.000000000 +0100
+++ gnutls-3.1.10/tests/openpgp-certs/testcerts	2013-03-26 15:52:33.207007952 +0100
@@ -23,7 +23,7 @@
 srcdir="${srcdir:-.}"
 SERV="${SERV:-../../src/gnutls-serv} -q"
 CLI="${CLI:-../../src/gnutls-cli}"
-PORT="${PORT:-5557}"
+PORT="${PORT:-$((RANDOM+5000))}"
 DEBUG=""
 
 if test "${WINDIR}" != "";then
