diff -up gnutls-2.12.7/lib/gcrypt/pk.c.mpi gnutls-2.12.7/lib/gcrypt/pk.c
--- gnutls-2.12.7/lib/gcrypt/pk.c.mpi	2011-04-08 02:30:44.000000000 +0200
+++ gnutls-2.12.7/lib/gcrypt/pk.c	2011-07-25 14:21:43.000000000 +0200
@@ -112,7 +112,7 @@ _wrap_gcry_pk_encrypt (gnutls_pk_algorit
       goto cleanup;
     }
 
-  res = gcry_sexp_nth_mpi (list, 1, 0);
+  res = gcry_sexp_nth_mpi (list, 1, GCRYMPI_FMT_USG);
   gcry_sexp_release (list);
   if (res == NULL)
     {
@@ -202,7 +202,7 @@ _wrap_gcry_pk_decrypt (gnutls_pk_algorit
       goto cleanup;
     }
 
-  res = gcry_sexp_nth_mpi (s_plain, 0, 0);
+  res = gcry_sexp_nth_mpi (s_plain, 0, GCRYMPI_FMT_USG);
   if (res == NULL)
     {
       gnutls_assert ();
@@ -327,7 +327,7 @@ _wrap_gcry_pk_sign (gnutls_pk_algorithm_
             goto cleanup;
           }
 
-        res[0] = gcry_sexp_nth_mpi (list, 1, 0);
+        res[0] = gcry_sexp_nth_mpi (list, 1, GCRYMPI_FMT_USG);
         gcry_sexp_release (list);
 
         list = gcry_sexp_find_token (s_sig, "s", 0);
@@ -338,7 +338,7 @@ _wrap_gcry_pk_sign (gnutls_pk_algorithm_
             goto cleanup;
           }
 
-        res[1] = gcry_sexp_nth_mpi (list, 1, 0);
+        res[1] = gcry_sexp_nth_mpi (list, 1, GCRYMPI_FMT_USG);
         gcry_sexp_release (list);
 
         ret = _gnutls_encode_ber_rs (signature, res[0], res[1]);
@@ -360,7 +360,7 @@ _wrap_gcry_pk_sign (gnutls_pk_algorithm_
             goto cleanup;
           }
 
-        res[0] = gcry_sexp_nth_mpi (list, 1, 0);
+        res[0] = gcry_sexp_nth_mpi (list, 1, GCRYMPI_FMT_USG);
         gcry_sexp_release (list);
 
         ret = _gnutls_mpi_dprint (res[0], signature);
@@ -559,7 +559,7 @@ _dsa_generate_params (bigint_t * resarr,
       return GNUTLS_E_INTERNAL_ERROR;
     }
 
-  resarr[0] = gcry_sexp_nth_mpi (list, 1, 0);
+  resarr[0] = gcry_sexp_nth_mpi (list, 1, GCRYMPI_FMT_USG);
   gcry_sexp_release (list);
 
   list = gcry_sexp_find_token (key, "q", 0);
@@ -570,7 +570,7 @@ _dsa_generate_params (bigint_t * resarr,
       return GNUTLS_E_INTERNAL_ERROR;
     }
 
-  resarr[1] = gcry_sexp_nth_mpi (list, 1, 0);
+  resarr[1] = gcry_sexp_nth_mpi (list, 1, GCRYMPI_FMT_USG);
   gcry_sexp_release (list);
 
   list = gcry_sexp_find_token (key, "g", 0);
@@ -581,7 +581,7 @@ _dsa_generate_params (bigint_t * resarr,
       return GNUTLS_E_INTERNAL_ERROR;
     }
 
-  resarr[2] = gcry_sexp_nth_mpi (list, 1, 0);
+  resarr[2] = gcry_sexp_nth_mpi (list, 1, GCRYMPI_FMT_USG);
   gcry_sexp_release (list);
 
   list = gcry_sexp_find_token (key, "y", 0);
@@ -592,7 +592,7 @@ _dsa_generate_params (bigint_t * resarr,
       return GNUTLS_E_INTERNAL_ERROR;
     }
 
-  resarr[3] = gcry_sexp_nth_mpi (list, 1, 0);
+  resarr[3] = gcry_sexp_nth_mpi (list, 1, GCRYMPI_FMT_USG);
   gcry_sexp_release (list);
 
 
@@ -604,7 +604,7 @@ _dsa_generate_params (bigint_t * resarr,
       return GNUTLS_E_INTERNAL_ERROR;
     }
 
-  resarr[4] = gcry_sexp_nth_mpi (list, 1, 0);
+  resarr[4] = gcry_sexp_nth_mpi (list, 1, GCRYMPI_FMT_USG);
 
   gcry_sexp_release (list);
   gcry_sexp_release (key);
@@ -660,7 +660,7 @@ _rsa_generate_params (bigint_t * resarr,
       return GNUTLS_E_INTERNAL_ERROR;
     }
 
-  resarr[0] = gcry_sexp_nth_mpi (list, 1, 0);
+  resarr[0] = gcry_sexp_nth_mpi (list, 1, GCRYMPI_FMT_USG);
   gcry_sexp_release (list);
 
   list = gcry_sexp_find_token (key, "e", 0);
@@ -671,7 +671,7 @@ _rsa_generate_params (bigint_t * resarr,
       return GNUTLS_E_INTERNAL_ERROR;
     }
 
-  resarr[1] = gcry_sexp_nth_mpi (list, 1, 0);
+  resarr[1] = gcry_sexp_nth_mpi (list, 1, GCRYMPI_FMT_USG);
   gcry_sexp_release (list);
 
   list = gcry_sexp_find_token (key, "d", 0);
@@ -682,7 +682,7 @@ _rsa_generate_params (bigint_t * resarr,
       return GNUTLS_E_INTERNAL_ERROR;
     }
 
-  resarr[2] = gcry_sexp_nth_mpi (list, 1, 0);
+  resarr[2] = gcry_sexp_nth_mpi (list, 1, GCRYMPI_FMT_USG);
   gcry_sexp_release (list);
 
   list = gcry_sexp_find_token (key, "p", 0);
@@ -693,7 +693,7 @@ _rsa_generate_params (bigint_t * resarr,
       return GNUTLS_E_INTERNAL_ERROR;
     }
 
-  resarr[3] = gcry_sexp_nth_mpi (list, 1, 0);
+  resarr[3] = gcry_sexp_nth_mpi (list, 1, GCRYMPI_FMT_USG);
   gcry_sexp_release (list);
 
 
@@ -705,7 +705,7 @@ _rsa_generate_params (bigint_t * resarr,
       return GNUTLS_E_INTERNAL_ERROR;
     }
 
-  resarr[4] = gcry_sexp_nth_mpi (list, 1, 0);
+  resarr[4] = gcry_sexp_nth_mpi (list, 1, GCRYMPI_FMT_USG);
   gcry_sexp_release (list);
 
 
@@ -717,7 +717,7 @@ _rsa_generate_params (bigint_t * resarr,
       return GNUTLS_E_INTERNAL_ERROR;
     }
 
-  resarr[5] = gcry_sexp_nth_mpi (list, 1, 0);
+  resarr[5] = gcry_sexp_nth_mpi (list, 1, GCRYMPI_FMT_USG);
 
   gcry_sexp_release (list);
   gcry_sexp_release (key);
